### ISSUE TITLE
Changing a field from private to public on a type implementing Deref can break downstream code.

### ISSUE THREAD

**Issue by theemathas** [obi1kenobi/cargo-semver-checks#1297]
### Is this about an existing lint, or proposing a new one?

new

### Known issues that might be causing this

- [ ] Is the item that should have been flagged originating from another crate, or from a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))
- [ ] Is the issue related to a change in the type of a field, function parameter, generic, or trait bound? ([#149](https://github.com/obi1kenobi/cargo-semver-checks/issues/149))
- [ ] If you're proposing a new lint, is it already part of our lint wishlist? ([#5](https://github.com/obi1kenobi/cargo-semver-checks/issues/5))

### Steps to reproduce the bug with the above code

Starting with the following code:

```rust
use std::ops::Deref;

pub struct Foo((i32, i64));

impl Foo {
    pub fn new(x: i32, y: i64) -> Self {
        Self((x, y))
    }
}

impl Deref for Foo {
    type Target = (i32, i64);

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
```

I edited the field of `Foo` to make it public.

<details>
<summary>Code with public field</summary>

```rust
use std::ops::Deref;

pub struct Foo(pub (i32, i64));

impl Foo {
    pub fn new(x: i32, y: i64) -> Self {
        Self((x, y))
    }
}

impl Deref for Foo {
    type Target = (i32, i64);

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
```

</details>

### Actual Behaviour

cargo-semver-checks does not report any semver issues

### Expected Behaviour

cargo-semver-checks should report a semver issue, since the following code would stop compiling in a downstream crate:

```rust
use foo::Foo;

pub fn check(x: Foo) -> i32 {
    x.0
}
```

#### Software version

cargo-semver-checks 0.41.0

#### Operating system

- OS: macOS 15.5 Sequoia
- Kernel: 24.5.0

#### Command-line

```bash
/Users/timch/.cargo/bin/cargo-semver-checks semver-checks --bugreport 
```

#### cargo version

```
> cargo -V 
cargo 1.87.0 (99624be96 2025-05-06)
```

#### Compile time information

- Profile: release
- Target triple: aarch64-apple-darwin
- Family: unix
- OS: macos
- Architecture: aarch64
- Pointer width: 64
- Endian: little
- CPU features: aes,crc,dit,dotprod,dpb,dpb2,fcma,fhm,flagm,fp16,frintts,jsconv,lor,lse,neon,paca,pacg,pan,pmuv3,ras,rcpc,rcpc2,rdm,sb,sha2,sha3,ssbs,vh
- Host: aarch64-apple-darwin

**Comment by obi1kenobi** [comment]
Nice one!

I think it's not just fields — any `pub` associated items (functions, consts, etc.) have this SemVer hazard too. If the target `Deref` type happens to have an associated item by the same name as the `impl Deref` type, we get the same issue: the inherent item takes precedence over the `Deref`'s item.

Even proving that _today_ the `Deref` target type doesn't have such a conflicting item isn't sufficient to prove the absence of a SemVer hazard. Unless that target type is defined by the same crate as the `impl Deref` type (i.e. they are forced to upgrade in lockstep), a future minor/patch release of the target type's crate can add items that would cause this breakage: `impl_deref v1.0 + deref_target v1.1` could resolve to `deref_target::Target::item` while `impl_deref v1.1-semver-break + deref_target v1.0` could reoslve to `impl_deref::DerefWrapper::item`. In this case, even if `impl_deref v1.1-semver-break` is released first, users that upgrade to `deref_target v1.1` _first_ and `impl_deref v1.1-semver-break` second will experience breakage.

So I think we have to treat new additions of _any_ `pub` associated items (fields, associated functions / methods, consts, maybe even variants — not sure about them) on an _existing_ `impl Deref` type as major breaking changes. There are probably 3-4 lints here!

Any interest in writing some of those lints? Most lints are [community-contributed](https://github.com/obi1kenobi/cargo-semver-checks/blob/main/CONTRIBUTING.md#adding-a-new-lint), and I'd be happy to help & point you in the right direction!

**Comment by Nadrieril** [comment]
> I think it's not just fields — any pub associated items (functions, consts, etc.) have this SemVer hazard too.

That's not exactly true: only items subject to postfix notation are concerned. Explicit `Type::method` notation does not go through autoderef and thus is not affected by potential shadowing (that's why e.g. all the `Box` methods take `this: Self` instead of `self`). So you're right that methods count too, but only methods that take a `&self`/`&mut self`/etc argument.

**Comment by obi1kenobi** [comment]
Nice!

The hardest part of implementing SemVer lints is figuring out the exact boundaries of the rule we need to implement, down to the fine details like you just described.

Pitching in to help with that research is super valuable and high leverage for the project, and I'm really grateful for it 🙏
